barretenberg/cpp/pil/avm/main.pil

```diff
@@ -479,46 +428,33 @@ namespace main(256);
-    //       are activated through the relation below. The other opcodes which are implemented purely
-    //       through memory sub-operations such as RETURN, SET are activated by
-    //       setting a newly introduced boolean sel_mem_op_activate_gas which is set in witness generation.
-    //       We should remove this shortcut and constrain this activation through bytecode decomposition.
-    //       Alternatively, we introduce a boolean selector for the three opcodes mentioned above.
-    //       Note: External call gas cost is not constrained
-    pol commit sel_gas_accounting_active;
-    // TODO: remove sload and sstore from here
-    // This temporarily disables gas tracking for sload and sstore because our gas
-    // tracking doesn't work properly for instructions that span multiple rows
-    // TODO: disabling this until PR in stack.
-    // sel_gas_accounting_active - OPCODE_SELECTORS - SEL_ALL_CTRL_FLOW - sel_op_sload - sel_op_sstore - sel_mem_op_activate_gas = 0;
-
-    // Program counter must increment if not jumping or returning
-    // TODO: support for muli-rows opcode in execution trace such as
-    // radix, hash gadgets operations. At the moment, we have to increment
-    // the pc in witness generation for all rows pertaining to the original
-    // opcode. This is misleading. Ultimately, we want the pc to be incremented
-    // just after the last row of a given opcode.
+    pol SEL_ALL_GADGET = sel_op_radix_le + sel_op_sha256 + sel_op_poseidon2 + sel_op_keccak + sel_op_pedersen
+                       + sel_op_ecadd + sel_op_pedersen_commit + sel_op_msm;
+    pol SEL_ALL_MEMORY = sel_op_cmov + sel_op_mov + sel_op_set;
+    pol OPCODE_SELECTORS = sel_op_fdiv + sel_op_calldata_copy + sel_op_get_contract_instance
+                         + SEL_ALL_ALU + SEL_ALL_BINARY + SEL_ALL_MEMORY + SEL_ALL_GADGET
+                         + KERNEL_INPUT_SELECTORS + KERNEL_OUTPUT_SELECTORS + SEL_ALL_LEFTGAS
+                         + SEL_ALL_CTRL_FLOW;
+
+    pol CUR_AND_NEXT_ARE_MAIN = sel_execution_row * sel_execution_row';
+
+    // When considering two adjacent main trace rows,
+    // the program counter must increment if not jumping or returning.
     #[PC_INCREMENT]
-    (1 - sel_first) * (1 - sel_op_halt) * OPCODE_SELECTORS * (pc' - (pc + 1)) = 0;
+    CUR_AND_NEXT_ARE_MAIN * (1 - SEL_ALL_CTRL_FLOW) * (pc' - (pc + 1)) = 0;
 
-    // sel_first == 0 && sel_op_internal_call == 0 && sel_op_internal_return == 0 && sel_op_halt == 0 ==> internal_return_ptr == internal_return_ptr'
+    // When considering two adjacent main trace rows,
+    // the internal return ptr must stay the same if not jumping or returning.
     #[INTERNAL_RETURN_POINTER_CONSISTENCY]
-    (1 - INTERNAL_CALL_STACK_SELECTORS) * (internal_return_ptr' - internal_return_ptr) = 0; 
+    CUR_AND_NEXT_ARE_MAIN * (1 - SEL_ALL_CTRL_FLOW) * (internal_return_ptr' - internal_return_ptr) = 0; 
```

Comment:
We should write a TODO that for external_call and external_return the internal_return_ptr will actually have to be reset (to zero for an external call) or resumed (for an external return) and that the space_id/call_ptr will be incremented. In any case, when we properly support external_return, the above relation will have to be changed.